src/components/Products/ProductsTable.tsx

```diff
@@ -0,0 +1,146 @@
+import TableRow from "@material-ui/core/TableRow";
+import { withStyles } from "@material-ui/core/styles";
+import { createStyles, Theme } from "@material-ui/core/styles";
+import { ProductsTableProps } from "../../types";
+import TableSortLabel from "@material-ui/core/TableSortLabel";
+import useSortTable from "../../customHooks/useSortTable";
+import useTablePagination from "../../customHooks/useTablePagination";
+import IconButton from "@material-ui/core/IconButton";
+import DeleteForeverIcon from "@material-ui/icons/DeleteForever";
+import EditIcon from "@material-ui/icons/Edit";
+import DescriptionIcon from "@material-ui/icons/Description";
+import DeleteDialog from "../Categories/DeleteDialog";
+
+const styles = (theme: Theme) =>
+  createStyles({
+    root: {
+      width: "95%",
+    },
+    container: {
+      maxHeight: 280,
+    },
+    tableCell: {
+      width: 120,
+      align: "left",
+      padding: theme.spacing(0.1, 0, 0.1, 2),
+    },
+  });
+
+const ProductsTable: React.FC<ProductsTableProps> = ({
+  classes,
+  columns = [],
+  rows: incomingRows = [],
+  onDelete: handleDelete,
+  onEdit: handleEdit,
+  onViewDetails: handleViewDetails,
+}) => {
+  const rows = useMemo(() => incomingRows.slice(), [incomingRows]);
+
+  const { createSortHandler, stableSort, order, orderBy } = useSortTable(
+    columns[0].id
+  );
+
+  const { tablePagination, page, rowsPerPage } = useTablePagination(rows);
+
+  const tableHead = (
+    <TableHead>
+      <TableRow>
+        {columns.map((column) => (
+          <TableCell
+            key={column.id}
+            align={column.align}
+            style={{ width: column.minWidth }}
+          >
+            <TableSortLabel
+              active={orderBy === column.id}
+              direction={orderBy === column.id ? order : "asc"}
+              onClick={createSortHandler(column.id)}
+            >
+              {column.label}
+            </TableSortLabel>
+          </TableCell>
+        ))}
+        <TableCell className={classes.tableCell}>{"Actions"}</TableCell>
+      </TableRow>
+    </TableHead>
+  );
+
+  if (rows.length === 0)
+    return (
+      <Paper className={classes.root}>
+        <TableContainer className={classes.container}>
+          <Table stickyHeader aria-label="sticky table">
+            {tableHead}
+            <TableBody>
+              <TableRow>
+                <TableCell className={classes.tableCell}>
+                  {"No matching records found"}
+                </TableCell>
+              </TableRow>
+            </TableBody>
+          </Table>
+        </TableContainer>
+      </Paper>
+    );
+
+  return (
+    <Paper className={classes.root}>
+      <TableContainer className={classes.container}>
+        <Table stickyHeader aria-label="sticky table">
+          {tableHead}
+          <TableBody>
+            {stableSort(rows, order, orderBy)
+              .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
```

Comment:
usually prefer to extract all logic stuff before the JSX i.e even the sort / slice be done before returning the JSX 
why ? so that we dont search for logic between dom 
2. we can utilize useMemo in top so we dont execute this logic on each render, but only if logic-dependent vars changes